src/HSE/All.hs

```diff
@@ -139,45 +155,53 @@ runCpp (Cpphs o) file x = dropLine <$> runCpphs o file x
+data ParsedModuleResults = ParsedModuleResults {
+    pm_hsext  :: (Module SrcSpanInfo, [Comment]) -- hs-src-ext result.
+  , pm_ghclib :: Located (HsSyn.HsModule HsSyn.GhcPs) -- ghc-lib-parser result.
+}
+
 -- | Parse a Haskell module. Applies the C pre processor, and uses best-guess fixity resolution if there are ambiguities.
---   The filename @-@ is treated as @stdin@. Requires some flags (often 'defaultParseFlags'), the filename, and optionally the contents of that file.
-parseModuleEx :: ParseFlags -> FilePath -> Maybe String -> IO (Either ParseError (Module SrcSpanInfo, [Comment]))
+-- The filename @-@ is treated as @stdin@. Requires some flags (often 'defaultParseFlags'), the filename, and optionally the contents of that file.
+-- This version uses both hs-src-exts AND ghc-lib. It's considered to be an unrecoverable error if one
+-- parsing method succeeds whilst the other fails.
+parseModuleEx :: ParseFlags -> FilePath -> Maybe String -> IO (Either ParseError ParsedModuleResults)
 parseModuleEx flags file str = timedIO "Parse" file $ do
         str <- case str of
             Just x -> return x
             Nothing | file == "-" -> getContentsUTF8
                     | otherwise -> readFileUTF8' file
         str <- return $ fromMaybe str $ stripPrefix "\65279" str -- remove the BOM if it exists, see #130
         ppstr <- runCpp (cppFlags flags) file str
-        case parseFileContentsWithComments (mode flags) ppstr of
-            ParseOk (x, cs) -> return $ Right (applyFixity fixity x, cs)
-            ParseFailed sl msg -> do
-                -- figure out the best line number to grab context from, by reparsing
-                -- but not generating {-# LINE #-} pragmas
-                flags <- return $ parseFlagsNoLocations flags
-                ppstr2 <- runCpp (cppFlags flags) file str
-                let pe = case parseFileContentsWithMode (mode flags) ppstr2 of
-                        ParseFailed sl2 _ -> context (srcLine sl2) ppstr2
-                        _ -> context (srcLine sl) ppstr
-                return $ Left $ ParseError sl msg pe
+        case (parseFileContentsWithComments (mode flags) ppstr, parseFileGhcLib file ppstr)
+          of
+            (ParseOk (x, cs), POk _ mod) ->
+              return $ Right (ParsedModuleResults (applyFixity fixity x, cs) mod)
+            (ParseFailed sl msg, PFailed ps) -> do
+              flags <- return $ parseFlagsNoLocations flags
+              ppstr2 <- runCpp (cppFlags flags) file str
+              let pe = case parseFileContentsWithMode (mode flags) ppstr2 of
+                         ParseFailed sl2 _ -> context (srcLine sl2) ppstr2
+                         _ -> context (srcLine sl) ppstr
+              return $ Left $ ParseError sl msg pe (pprErrMsgBagWithLoc $ snd $ getMessages ps dynFlags)
+            (ParseOk _, PFailed _) ->
+              error "Unexpected : hs-src-exts succeeded, ghc-lib-parser failed"
```

Comment:
Still not keen on this. As soon as it gets released, I'll get a million bug reports with this error and the one below.

Reply:
Hmm. Ok. Ideas on what to do?